src/engine.ts

```diff
@@ -467,24 +467,60 @@ export default class Engine {
   }
   
   cleanUpPhase() {
-  
+
     for (const player of this.playersInOrder()) {
       // remove roundScoringTile
-      player.removeEvents( Event.parse(roundScorings[this.roundScoringTiles[this.round]]));
+      player.removeEvents(Event.parse(roundScorings[this.roundScoringTiles[this.round]]));
 
       // resets special action
       for (const event of player.events[Operator.Activate]) {
         event.activated = false;
       }
     }
     // resets power and qic actions
-    Object.values(BoardAction).forEach( pos => {
+    Object.values(BoardAction).forEach(pos => {
       this.boardActions[pos] = true;
     });
+  }
+
+  finalScoringPhase() {
+    //finalScoring tiles
+    for (const tile of this.finalScoringTiles) {
+      const players = _.sortBy(this.players, player => player.finalCount(tile)).reverse();
+
+      const rankings = players.map(pl => ({
+        player: pl,
+        count: pl.finalCount(tile)
+      }));
+
+      if (this.players.length === 2) {
+        rankings.push({ player: null, count: 8 });
+        rankings.sort((pl1, pl2) => pl2.count - pl1.count);
+      }
 
+      for (const ranking of rankings) {
+        const count = ranking.count;
+        // index of the first player with that score
+        const first = rankings.findIndex(pl => pl.count === count);
+        // number of other players with the same score
+        const ties = rankings.filter(pl => pl.count === count).length;
 
+        if (ranking.player) {
+          const VPs = [18, 12, 6, 0, 0, 0];
+          ranking.player.data.victoryPoints += Math.floor(_.sum(VPs.slice(first, ties)) / ties);
+        }
+      }
+    }
+
+    //research VP and remaining resources
+    for (const pl of this.playersInOrder()) {
+      Object.values(ResearchField).forEach(research => (pl.data.victoryPoints += Math.max(pl.data.research[research] - 3, 0) * 4));
+      const resources = pl.data.ores + pl.data.credits + pl.data.qics + pl.data.knowledge + (Math.ceil(pl.data.power.area2 / 2) + pl.data.power.area3);
```

Comment:
It should be `Math.floor`, not `Math.ceil`

Reply:
Also, I feel that this should be included in a function like `Player.gainFinalVictoryPoints()`. Because the code is going to be more complicated when dealing with factions like Taklons.

Reply:
Let's refactor it when we finalise Taklons